dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/common/CteDefinition.java

```diff
@@ -29,54 +29,260 @@
+        queryItemId,
+        programStageUid,
+        null,
+        cteDefinition,
+        generateRandomAlias(),
+        false, // rowContext
+        false, // isExists
+        false, // requiresCoalesce
+        null, // aggregateWhereClause
+        null, // joinColumn
+        null, // targetRank
+        CteType.PROGRAM_STAGE); // Set type
+    this.offsets.add(offset);
+  }
 
-  /** Whether the CTE is a row context * */
-  @Getter private boolean rowContext;
+  // Constructor for standard Program Stage CTEs with rowContext
+  public CteDefinition(
+      String programStageUid,
+      String queryItemId,
+      String cteDefinition,
+      int offset,
+      boolean isRowContext) {
+    this(
+        queryItemId,
+        programStageUid,
+        null,
+        cteDefinition,
+        generateRandomAlias(),
+        isRowContext, // Pass rowContext
+        false, // isExists
+        false, // requiresCoalesce
+        null, // aggregateWhereClause
+        null, // joinColumn
+        null, // targetRank
+        CteType.PROGRAM_STAGE); // Set type
+    this.offsets.add(offset);
+  }
 
-  /** Whether the CTE is a program indicator * */
-  @Getter private boolean programIndicator = false;
+  public CteDefinition(String cteDefinition, String aggregateWhereClause) {
+    this(
+        ENROLLMENT_AGGR_BASE, // itemId
+        null, // programStageUid
+        null, // programIndicatorUid
+        cteDefinition,
+        generateRandomAlias(), // alias
+        false, // rowContext
+        false, // isExists
+        false, // requiresCoalesce
+        aggregateWhereClause, // Pass aggregateWhereClause
+        null, // joinColumn
+        null, // targetRank
+        CteType.BASE_AGGREGATION); // Set type
+  }
 
-  /** Whether the CTE is a filter * */
-  @Getter private boolean filter = false;
+  /** Creates a CTE definition for program indicators. */
+  public static CteDefinition forProgramIndicator(
+      String programIndicatorUid, String cteDefinition, boolean requiresCoalesce) {
+    // Calls private constructor, passing CteType.PROGRAM_INDICATOR
+    return new CteDefinition(
+        programIndicatorUid, // itemId
+        null, // programStageUid
+        programIndicatorUid, // programIndicatorUid
+        cteDefinition,
+        generateRandomAlias(), // alias
+        false, // rowContext
+        false, // isExists
+        requiresCoalesce, // requiresCoalesce
+        null, // aggregateWhereClause
+        null, // joinColumn
+        null, // targetRank
+        CteType.PROGRAM_INDICATOR); // Set type
+  }
 
-  /** Whether the CTE is a exists, used for checking if the enrollment exists * */
-  private boolean isExists = false;
+  /** Creates a CTE definition for filter CTEs (replacing filter subqueries). */
+  public static CteDefinition forFilter(
+      String queryItemId, String programStageUid, String cteDefinition) {
+    // Calls private constructor, passing CteType.FILTER
+    return new CteDefinition(
```

Comment:
Not saying it necessarily is better, though perhaps this is a sign that we should use a builder class instead. We can get one from Lombok. A high number of null and false parameters could lead to issues and maintenance problelsm later.